### PR TITLE
Asynchronous Groth16 verification

### LINKED ISSUE #305: Implement verifiers that accept Groth16 types
Require bellman or bellman community, validators can accept our Groth16 types and validate them, with batching or without.

https://github.com/ZcashFoundation/zebra/blob/main/zebra-chain/src/proofs/groth16.rs

image

https://zips.z.cash/protocol/protocol.pdf#groth

### LINKED ISSUE #406: Implement Groth16 batch verification optimizations
https://zips.z.cash/protocol/protocol.pdf#grothbatchverify

This impl can be swapped out for calling out directly to Bellman when complete.

image
image

### PR DESCRIPTION
**Pull request by hdevalence**
This PR is the first step in getting a groth16 proving system fully integrated with the rest of zebra. This PR implements the initial async API, but none of the actual batching logic necessary for our eventual verifier design.

Once the batch verification API from bellman has been implemented we will need to swap out the "Batch" type defined in this crate with the new `batch::Verifier` defined in bellman.

## Follow Up Work

- https://github.com/ZcashFoundation/zebra/issues/1645
- https://github.com/zkcrypto/bellman/pull/59

### CODE REVIEW

File: zebra-consensus/src/primitives/groth16.rs
Comment (hdevalence) [inline zebra-consensus/src/primitives/groth16.rs]: This prevents the `E: ... ` bounds from propagating through `zebra-consensus`.

**Comment by hdevalence** [comment]
Rebased onto `main` with changes from #821

**Comment by hdevalence** [comment]
> Using a WIP branch on the ZF fork of librustzcash for now. Unsure whether it's better to create an intermediate step of having the batch verification API do single verification internally or just go directly to what we want, but I'm leaning towards the latter, since we already know (from the Ed25519 and Redjubjub cases) that an async API with this shape will work.

Now leaning towards a middle option, to create a `fake-batch` branch that uses the batch verification interface currently defined but internally implements batch verification by calling single verification N times.  This lets us reuse the interface, the item types, etc., and we can swap out the math later when we have more time.

File: zebra-consensus/src/primitives/groth16.rs
Comment (dconnolly) [reply zebra-consensus/src/primitives/groth16.rs]: thank goodness 😅

**Comment by dconnolly** [comment]
It builds!

**Comment by dconnolly** [comment]
@yaahc I moved to 'ready for review' to get the other CI jobs running, you can move back if you want EDIT: i can trigger those jobs without moving from draft

**Comment by dconnolly** [comment]
@yaahc I rebased because of a conflict related to a bls12 dependency update

File: zebra-consensus/src/primitives.rs
Comment (teor2345) [inline zebra-consensus/src/primitives.rs]:
```suggestion
/// The size of the buffer in the broadcast channels used by batch verifiers.
///
/// This bound limits the number of concurrent batches for each verifier.
/// If tasks delay checking for verifier results, and the bound is too small,
/// new batches will be rejected with `RecvError`s.
const BROADCAST_BUFFER_SIZE: usize = 512;
```

File: zebra-consensus/src/primitives/groth16.rs
Comment (teor2345) [inline zebra-consensus/src/primitives/groth16.rs]: You might want to delete the commented-out type alias, and use the doc comment for the current `Item` type:
```suggestion
/// A Groth16 verification item, used as the request type of the service.
#[derive(Clone)]
pub struct Item<E: MultiMillerLoop> {
```

File: zebra-consensus/src/primitives/groth16.rs
Comment (teor2345) [inline zebra-consensus/src/primitives/groth16.rs]: I'm not sure what this PR is intended to do, so I don't know if the commented-out code in this section is:
* fields that will be added in future, or
* old code that should be deleted.

Or a combination of both.

**Review by teor2345 — DISMISSED** [review]
As far as I understand, this code looks good.

But it's hard to review without:
* a PR description that explains the current purpose of this PR
* doc comments on functions and types
* tests

File: zebra-consensus/src/primitives/groth16.rs
Comment (dconnolly) [inline zebra-consensus/src/primitives/groth16.rs]: Can we remove these commented-out debug fields?

File: zebra-consensus/src/primitives/groth16.rs
Comment (yaahc) [reply zebra-consensus/src/primitives/groth16.rs]: it was the code I wanted to write, but couldn't without a PR to bellman, which should be a part of the PR that deirdre is working on, which I think does `#[derive(Debug)]` on PreparedVerifyingKey.

File: zebra-consensus/src/primitives/groth16.rs
Comment (yaahc) [reply zebra-consensus/src/primitives/groth16.rs]: The type alias should stay, since the `Item` is a filler for the item type from `bellman`, once the batch API has been completed, so all the comments left are the code that is supposed to be ripped out for the real implementation once the crypto side of things is finished. I'll go ahead and move all the temporary filler types to one part of the file and document why they're there and how they should be handled, but I'd like to leave in these commented lines because they should already be correct, and only need to be uncommented once we swap to the new bellman.

Comment (teor2345) [reply zebra-consensus/src/primitives/groth16.rs]: Sounds good.

Sometimes there's a bit of a delay between follow-up PRs, so it's good to have a comment that references the follow-up.

File: zebra-consensus/src/script.rs
Comment (teor2345) [inline zebra-consensus/src/script.rs]: This change reverts #1680, but #1680 is correct

File: zebra-consensus/src/script.rs
Comment (teor2345) [inline zebra-consensus/src/script.rs]: We should keep this comment, rather than deleting it

**Comment by teor2345** [comment]
If we're planning multiple PRs that change the same code, let's leave the doc comments until the later PRs, so we don't accidentally delete or revert any more comments.

File: zebra-consensus/src/script.rs
Comment (yaahc) [reply zebra-consensus/src/script.rs]: oh yea, unintentional. I "squashed" my commits by doing `git reset --soft main` after merging main in and it looks like #1680 merged after so im guessing it just tripped up the merge resolution because of the times? Not really sure how it broke but yea, fixing

**Comment by yaahc** [comment]
> If we're planning multiple PRs that change the same code, let's leave the doc comments until the later PRs, so we don't accidentally delete or revert any more comments.

@teor2345 is this resolved btw? I assumed you meant "lets leave adding the doc comments until the later PR" but it never got approved after I did the rest of the cleanup so now I'm unsure if there's still something left for me to do on this PR.

**Review by dconnolly — COMMENTED** [review]
I wonder if we should merge this now, knowing that when https://github.com/zkcrypto/bellman/pull/59 is merged in there will be changes anyway, or try to support [this pattern](https://github.com/ZcashFoundation/zebra/blob/bfb3de7a8a9c612774dd82d27f947f2c2dbdeaab/zebra-consensus/src/primitives/redjubjub.rs#L31) now

**Comment by yaahc** [comment]
> I wonder if we should merge this now, knowing that when [zkcrypto/bellman#59](https://github.com/zkcrypto/bellman/pull/59) is merged in there will be changes anyway, or try to support [this pattern](https://github.com/ZcashFoundation/zebra/blob/bfb3de7a8a9c612774dd82d27f947f2c2dbdeaab/zebra-consensus/src/primitives/redjubjub.rs#L31) now

I think merge personally, the PR is already setup to make it easy for the follow up bellman stuff, I don't see any reason to block this PR on it. As for the static verifiers, I think it might be good to add those with the PR that also adds the PreparedVerifyingKeys, since each static will need to hold a reference to a different key.

**Review by dconnolly — APPROVED** [review]

**Comment by dconnolly** [comment]
Merging, tracking followup work in #1645

**Comment by teor2345** [comment]
> > If we're planning multiple PRs that change the same code, let's leave the doc comments until the later PRs, so we don't accidentally delete or revert any more comments.
> 
> @teor2345 is this resolved btw? I assumed you meant "lets leave adding the doc comments until the later PR" but it never got approved after I did the rest of the cleanup so now I'm unsure if there's still something left for me to do on this PR.

I've been focused on other work.

Last time I checked this PR, it was unclear how much work was remaining, so I couldn't really approve it. So the best I could do was dismiss my review so I wasn't blocking it.

**Pull request merged**
